### PR TITLE
Add initial version of the General Diagnostics Cluster

### PR DESCRIPTION
**Pull request by yufengwangca**
#### Problem
The General Diagnostics Cluster, along with other diagnostics clusters, provide a means to acquire standardized diagnostics metrics that MAY be used by a Node to assist a user or Administrative Node in diagnosing potential problems.

The specification of General Diagnostics Cluster:
https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/master/src/service_device_management/DiagnosticsGeneral.adoc

 #### Summary of Changes
Add initial version of the General Diagnostics Cluster

### CODE REVIEW

**Review by andy31415 — APPROVED** [review]
I assume this is mostly generated code - PR is too large to review

**Review by woody-apple — APPROVED** [review]

**Comment by woody-apple** [comment]
@yufengwangca Sorry, conflict here :(

**Review by msandstedt — APPROVED** [review]

**Review by jelderton — APPROVED** [review]

File: src/app/zap-templates/templates/app/attribute-size-src.zapt [hunk 1]
```diff
@@ -86,7 +86,7 @@ uint16_t emberAfCopyList(ClusterId clusterId, EmberAfAttributeMetadata * am, boo
                 {{chipType}} * entry = reinterpret_cast<{{chipType}} *>(write ? src : dest);
                 {{#chip_attribute_list_entryTypes}}
                 {{#if (isString type)}}
-                chip::ByteSpan * {{name}}Span = &entry->{{name}}; // {{type}}
+                chip::ByteSpan * {{name}}Span = reinterpret_cast<chip::ByteSpan *>(&entry->{{name}}); // {{type}}
```
Comment (andy31415) [inline src/app/zap-templates/templates/app/attribute-size-src.zapt:89]: why are we reinterpret-cast here? why is the entry name not of the correct type?

Comment (bzbarsky-apple) [reply src/app/zap-templates/templates/app/attribute-size-src.zapt:89]: That's what https://github.com/project-chip/connectedhomeip/issues/6801 was about.

**Pull request merged**